### PR TITLE
Atomic compare_exchange implementation

### PR DESCRIPTION
**Pull request by adarshyoga**
- [x] Have you provided a meaningful PR description?

This PR contains the implementation of compare_exchange_strong and compare_exchange_weak atomic functions. It contains a test case that verifies both success and failure scenarios.

- [x] Have you added a test, reproducer or referred to an issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [ ] If this PR is a work in progress, are you filing the PR as a draft?

### CODE REVIEW

File: numba_dpex/kernel_api/atomic_ref.py [hunk 1]
```diff
@@ -200,3 +200,55 @@ def exchange(self, val):
         old = self._ref[self._index]
         self._ref[self._index] = val
         return old
+
+    def compare_exchange_weak(self, expected, desired, expected_idx=0):
```
Comment (diptorupd) [inline numba_dpex/kernel_api/atomic_ref.py:204]: I get the rationale, and having discussed it with you I is clear to me that `expected` needs to be an Array type. However, I still feel that from a general usage purpose it is rather confusing. We should create a `Ref` type in numba-dpex to make the usage clear.

File: numba_dpex/kernel_api/atomic_ref.py [hunk 2]
```diff
@@ -200,3 +200,55 @@ def exchange(self, val):
...
+        """Compares the value of the object referenced by the AtomicRef
+        against the value of ``expected[expected_idx]``.
+        If the values are equal, attempts to replace the value of the
+        referenced object with the value of ``desired``.
+        Otherwise assigns the original value of the
+        referenced object to ``expected[expected_idx]``.
+
+        Args:
+            expected : Vector containing the expected value of the
```
Comment (diptorupd) [inline numba_dpex/kernel_api/atomic_ref.py:213]: Python has no type called `Vector`. Perhaps array is the right term. But, again the usage will be confusing as to why someone needs to create an array to essentially pass a single element.

File: numba_dpex/kernel_api/atomic_ref.py [hunk 3]
```diff
@@ -200,3 +200,55 @@ def exchange(self, val):
...
+        replacement operation were successful.
+
+        """
+        if self._ref[self._index] == expected[expected_idx]:
+            self._ref[self._index] = desired
+            return True
+        expected[expected_idx] = self._ref[self._index]
+        return False
+
+    def compare_exchange_strong(self, expected, desired, expected_idx=0):
```
Comment (diptorupd) [inline numba_dpex/kernel_api/atomic_ref.py:230]: What is the difference between the two variants of `CAS`? When should one use one over the other?

Comment (diptorupd) [reply numba_dpex/kernel_api/atomic_ref.py:230]: Perhaps we should just have one variant?

File: numba_dpex/kernel_api/atomic_ref.py (hunk 1 above)
Comment (adarshyoga) [reply numba_dpex/kernel_api/atomic_ref.py]: Created #1316 for tracking work on new array ref type, which would be handled in a separate PR.

File: numba_dpex/experimental/_kernel_dpcpp_spirv_overloads/_atomic_ref_overloads.py [hunk 4]
```diff
@@ -323,6 +324,108 @@ def _intrinsic_exchange_gen(context, builder, sig, args):
...
+        desired_arg = args[2]
+        atomic_ref_ptr = builder.extract_value(
+            args[0],
+            context.data_model_manager.lookup(sig.args[0]).get_field_position(
+                "ref"
+            ),
+        )
+        # add conditional bitcast for atomic_ref pointer,
+        # expected[expected_idx], and desired
+        if sig.args[0].dtype == types.float32:
```
Comment (diptorupd) [inline numba_dpex/experimental/_kernel_dpcpp_spirv_overloads/_atomic_ref_overloads.py:364]: We can do it like follows as well to reduce code duplication:

```python
if isinstance(sig.args[0].dtype, types.Float):
    bitwidth = sig.args[0].dtype.bitwidth
    atomic_ref_ptr = builder.bitcast(
                atomic_ref_ptr,
                llvmir.PointerType(
                    llvmir.IntType(bitwidth), addrspace=sig.args[0].address_space
                ),
            )
            expected_arg = builder.bitcast(expected_arg, llvmir.IntType(bitwidth))
            desired_arg = builder.bitcast(desired_arg, llvmir.IntType(bitwidth))
```
Then we do not need two separate cases for fp32 and fp64.

File: numba_dpex/experimental/_kernel_dpcpp_spirv_overloads/_atomic_ref_overloads.py [hunk 5]
```diff
@@ -323,6 +324,108 @@ def _intrinsic_exchange_gen(context, builder, sig, args):
...
+        # If the return value is not same as expected, then store return
+        # value in expected.
+        # In either case, return result of cmp instruction.
+        is_cmp_exchg_success = builder.icmp_signed("==", ret_val, expected_arg)
+
+        with builder.if_else(is_cmp_exchg_success) as (then, otherwise):
+            with then:
+                pass
+            with otherwise:
+                if sig.args[0].dtype == types.float32:
```
Comment (diptorupd) [inline numba_dpex/experimental/_kernel_dpcpp_spirv_overloads/_atomic_ref_overloads.py:419]: Same here.

**Review by diptorupd — APPROVED** [review]
Overall looks good. I have one last improvement suggestion and after that the PR is good to go. Thank you @adarshyoga!

File: numba_dpex/experimental/_kernel_dpcpp_spirv_overloads/_atomic_ref_overloads.py (hunk 5 above)
Comment (diptorupd) [reply numba_dpex/experimental/_kernel_dpcpp_spirv_overloads/_atomic_ref_overloads.py:419]: No, cannot be done here as llvmlite.ir does not expose a single fp type.

File: numba_dpex/experimental/_kernel_dpcpp_spirv_overloads/_atomic_ref_overloads.py (hunk 4 above)
Comment (diptorupd) [reply numba_dpex/experimental/_kernel_dpcpp_spirv_overloads/_atomic_ref_overloads.py:364]: Let us skip as we have the same pattern in multiple places and let us just be consistent.

**Pull request merged**